File: src/bun.js/BuildMessage.zig
pub const BuildMessage = struct {
    pub const js = jsc.Codegen.JSBuildMessage;
    pub const toJS = js.toJS;
    pub const fromJS = js.fromJS;
    pub const fromJSDirect = js.fromJSDirect;

    msg: logger.Msg,
    // resolve_result: Resolver.Result,
    allocator: std.mem.Allocator,
    logged: bool = false,

    pub fn constructor(globalThis: *jsc.JSGlobalObject, _: *jsc.CallFrame) bun.JSError!*BuildMessage {
        return globalThis.throw("BuildMessage is not constructable", .{});
    }

    pub fn getNotes(this: *BuildMessage, globalThis: *jsc.JSGlobalObject) bun.JSError!jsc.JSValue {
        const notes = this.msg.notes;
        const array = try jsc.JSValue.createEmptyArray(globalThis, notes.len);
        for (notes, 0..) |note, i| {
            const cloned = try note.clone(bun.default_allocator);
            try array.putIndex(
                globalThis,
                @intCast(i),
                try BuildMessage.create(globalThis, bun.default_allocator, logger.Msg{ .data = cloned, .kind = .note }),
            );
        }

        return array;
    }

    pub fn toStringFn(this: *BuildMessage, globalThis: *jsc.JSGlobalObject) jsc.JSValue {
        const text = std.fmt.allocPrint(default_allocator, "BuildMessage: {s}", .{this.msg.data.text}) catch {
            return globalThis.throwOutOfMemoryValue();
        };
        var str = ZigString.init(text);
        str.setOutputEncoding();
        if (str.isUTF8()) {
            const out = str.toJS(globalThis);
            default_allocator.free(text);
            return out;
        }

        return str.toExternalValue(globalThis);
    }

    pub fn create(
        globalThis: *jsc.JSGlobalObject,
        allocator: std.mem.Allocator,
        msg: logger.Msg,
        // resolve_result: *const Resolver.Result,
    ) bun.OOM!jsc.JSValue {
        var build_error = try allocator.create(BuildMessage);
        build_error.* = BuildMessage{
            .msg = try msg.clone(allocator),
            // .resolve_result = resolve_result.*,
            .allocator = allocator,
        };

        return build_error.toJS(globalThis);
    }

    pub fn toString(
        this: *BuildMessage,
        globalThis: *jsc.JSGlobalObject,
        _: *jsc.CallFrame,
    ) bun.JSError!jsc.JSValue {
        return this.toStringFn(globalThis);
    }

    pub fn toPrimitive(
        this: *BuildMessage,
        globalThis: *jsc.JSGlobalObject,
        callframe: *jsc.CallFrame,
    ) bun.JSError!jsc.JSValue {
        const args_ = callframe.arguments_old(1);
        const args = args_.ptr[0..args_.len];
        if (args.len > 0) {
            if (!args[0].isString()) {
                return jsc.JSValue.jsNull();
            }

            const str = try args[0].getZigString(globalThis);
            if (str.eqlComptime("default") or str.eqlComptime("string")) {
                return this.toStringFn(globalThis);
            }
        }

        return jsc.JSValue.jsNull();
    }

    pub fn toJSON(
        this: *BuildMessage,
        globalThis: *jsc.JSGlobalObject,
        _: *jsc.CallFrame,
    ) bun.JSError!jsc.JSValue {
        var object = jsc.JSValue.createEmptyObject(globalThis, 4);
        object.put(globalThis, ZigString.static("name"), try bun.String.static("BuildMessage").toJS(globalThis));
        object.put(globalThis, ZigString.static("position"), this.getPosition(globalThis));
        object.put(globalThis, ZigString.static("message"), this.getMessage(globalThis));
        object.put(globalThis, ZigString.static("level"), this.getLevel(globalThis));
        return object;
    }

    pub fn generatePositionObject(msg: logger.Msg, globalThis: *jsc.JSGlobalObject) jsc.JSValue {
        const location = msg.data.location orelse return jsc.JSValue.jsNull();
        var object = jsc.JSValue.createEmptyObject(globalThis, 7);

        object.put(
            globalThis,
            ZigString.static("lineText"),
            ZigString.init(location.line_text orelse "").toJS(globalThis),
        );
        object.put(
            globalThis,
            ZigString.static("file"),
            ZigString.init(location.file).toJS(globalThis),
        );
        object.put(
            globalThis,
            ZigString.static("namespace"),
            ZigString.init(location.namespace).toJS(globalThis),
        );
        object.put(
            globalThis,
            ZigString.static("line"),
            JSValue.jsNumber(location.line),
        );
        object.put(
            globalThis,
            ZigString.static("column"),
            JSValue.jsNumber(location.column),
        );
        object.put(
            globalThis,
            ZigString.static("length"),
            JSValue.jsNumber(location.length),
        );
        object.put(
            globalThis,
            ZigString.static("offset"),
            JSValue.jsNumber(location.offset),
        );

        return object;
    }

    // https://github.com/oven-sh/bun/issues/2375#issuecomment-2121530202
    pub fn getColumn(this: *BuildMessage, _: *jsc.JSGlobalObject) jsc.JSValue {
        if (this.msg.data.location) |location| {
            return jsc.JSValue.jsNumber(@max(location.column - 1, 0));
        }

        return jsc.JSValue.jsNumber(@as(i32, 0));
    }

    pub fn getLine(this: *BuildMessage, _: *jsc.JSGlobalObject) jsc.JSValue {
        if (this.msg.data.location) |location| {
            return jsc.JSValue.jsNumber(@max(location.line - 1, 0));
        }

        return jsc.JSValue.jsNumber(@as(i32, 0));
    }

    pub fn getPosition(
        this: *BuildMessage,
        globalThis: *jsc.JSGlobalObject,
    ) jsc.JSValue {
        return BuildMessage.generatePositionObject(this.msg, globalThis);
    }

    pub fn getMessage(
        this: *BuildMessage,
        globalThis: *jsc.JSGlobalObject,
    ) jsc.JSValue {
        return ZigString.init(this.msg.data.text).toJS(globalThis);
    }

    pub fn getLevel(
        this: *BuildMessage,
        globalThis: *jsc.JSGlobalObject,
    ) jsc.JSValue {
        return ZigString.init(this.msg.kind.string()).toJS(globalThis);
    }

    pub fn finalize(this: *BuildMessage) void {
        this.msg.deinit(bun.default_allocator);
    }
};

const string = []const u8;

const std = @import("std");
const Resolver = @import("../resolver//resolver.zig").Resolver;

const bun = @import("bun");
const default_allocator = bun.default_allocator;
const logger = bun.logger;

const jsc = bun.jsc;
const JSGlobalObject = jsc.JSGlobalObject;
const JSValue = jsc.JSValue;
const ZigString = jsc.ZigString;

File: src/jsc/BuildMessage.rs
use std::io::Write as _;

use bun_alloc::AllocError;
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};
use bun_logger as logger;
use bun_str::ZigString;

#[bun_jsc::JsClass] // codegen: JSBuildMessage (toJS / fromJS / fromJSDirect wired by derive)
pub struct BuildMessage {
    pub msg: logger::Msg,
    // resolve_result: Resolver.Result,
    // PORT NOTE: `allocator: std.mem.Allocator` field dropped — global mimalloc.
    pub logged: bool,
}

impl Default for BuildMessage {
    fn default() -> Self {
        Self { msg: logger::Msg::default(), logged: false }
    }
}

impl BuildMessage {
    #[bun_jsc::host_fn]
    pub fn constructor(global: &JSGlobalObject, _frame: &CallFrame) -> JsResult<*mut BuildMessage> {
        global.throw("BuildMessage is not constructable", format_args!(""))
    }

    #[bun_jsc::host_fn(getter)]
    pub fn get_notes(&self, global: &JSGlobalObject) -> JsResult<JSValue> {
        let notes = &self.msg.notes;
        let array = JSValue::create_empty_array(global, notes.len())?;
        for (i, note) in notes.iter().enumerate() {
            let cloned = note.clone()?;
            array.put_index(
                global,
                u32::try_from(i).unwrap(),
                BuildMessage::create(global, logger::Msg { data: cloned, kind: logger::Kind::Note, ..Default::default() })?,
            )?;
        }

        Ok(array)
    }

    pub fn to_string_fn(&mut self, global: &JSGlobalObject) -> JSValue {
        // std.fmt.allocPrint → write! into Vec<u8>; Rust aborts on OOM so the
        // `catch { throwOutOfMemoryValue }` branch is unreachable here.
        let mut text: Vec<u8> = Vec::new();
        write!(&mut text, "BuildMessage: {}", bstr::BStr::new(&self.msg.data.text)).unwrap();

        let mut str = ZigString::init(&text);
        str.set_output_encoding();
        if str.is_utf8() {
            let out = str.to_js(global);
            // default_allocator.free(text) → `text` drops at scope end.
            return out;
        }

        // toExternalValue: JSC takes ownership of the backing buffer and frees it later.
        // TODO(port): ZigString::to_external_value must adopt `text`'s allocation; leak it here so
        // the external-string finalizer (mimalloc-backed) can free it.
        let leaked: &'static mut [u8] = Box::leak(text.into_boxed_slice());
        let mut str = ZigString::init(leaked);
        str.set_output_encoding();
        str.to_external_value(global)
    }

    pub fn create(
        global: &JSGlobalObject,
        msg: logger::Msg,
        // resolve_result: *const Resolver.Result,
    ) -> Result<JSValue, AllocError> {
        let build_error = Box::new(BuildMessage {
            msg: msg.clone()?,
            // resolve_result: resolve_result.*,
            logged: false,
        });

        Ok(build_error.to_js(global))
    }

    #[bun_jsc::host_fn(method)]
    pub fn to_string(
        &mut self,
        global: &JSGlobalObject,
        _frame: &CallFrame,
    ) -> JsResult<JSValue> {
        Ok(self.to_string_fn(global))
    }

    #[bun_jsc::host_fn(method)]
    pub fn to_primitive(
        &mut self,
        global: &JSGlobalObject,
        callframe: &CallFrame,
    ) -> JsResult<JSValue> {
        let args_ = callframe.arguments_old(1);
        let args = &args_.ptr[0..args_.len];
        if !args.is_empty() {
            if !args[0].is_string() {
                return Ok(JSValue::NULL);
            }

            let str = args[0].get_zig_string(global)?;
            if str.eql_comptime(b"default") || str.eql_comptime(b"string") {
                return Ok(self.to_string_fn(global));
            }
        }

        Ok(JSValue::NULL)
    }

    #[bun_jsc::host_fn(method)]
    pub fn to_json(
        &mut self,
        global: &JSGlobalObject,
        _frame: &CallFrame,
    ) -> JsResult<JSValue> {
        let object = JSValue::create_empty_object(global, 4);
        object.put(global, ZigString::static_str(b"name"), bun_str::String::static_str(b"BuildMessage").to_js(global)?);
        object.put(global, ZigString::static_str(b"position"), self.get_position(global));
        object.put(global, ZigString::static_str(b"message"), self.get_message(global));
        object.put(global, ZigString::static_str(b"level"), self.get_level(global));
        Ok(object)
    }

    pub fn generate_position_object(msg: &logger::Msg, global: &JSGlobalObject) -> JSValue {
        let Some(location) = &msg.data.location else { return JSValue::NULL; };
        let object = JSValue::create_empty_object(global, 7);

        object.put(
            global,
            ZigString::static_str(b"lineText"),
            ZigString::init(location.line_text.as_deref().unwrap_or(b"")).to_js(global),
        );
        object.put(
            global,
            ZigString::static_str(b"file"),
            ZigString::init(&location.file).to_js(global),
        );
        object.put(
            global,
            ZigString::static_str(b"namespace"),
            ZigString::init(&location.namespace).to_js(global),
        );
        object.put(
            global,
            ZigString::static_str(b"line"),
            JSValue::js_number(location.line),
        );
        object.put(
            global,
            ZigString::static_str(b"column"),
            JSValue::js_number(location.column),
        );
        object.put(
            global,
            ZigString::static_str(b"length"),
            JSValue::js_number(location.length),
        );
        object.put(
            global,
            ZigString::static_str(b"offset"),
            JSValue::js_number(location.offset),
        );

        object
    }

    // https://github.com/oven-sh/bun/issues/2375#issuecomment-2121530202
    #[bun_jsc::host_fn(getter)]
    pub fn get_column(&self, _global: &JSGlobalObject) -> JSValue {
        if let Some(location) = &self.msg.data.location {
            return JSValue::js_number((location.column - 1).max(0));
        }

        JSValue::js_number(0i32)
    }

    #[bun_jsc::host_fn(getter)]
    pub fn get_line(&self, _global: &JSGlobalObject) -> JSValue {
        if let Some(location) = &self.msg.data.location {
            return JSValue::js_number((location.line - 1).max(0));
        }

        JSValue::js_number(0i32)
    }

    #[bun_jsc::host_fn(getter)]
    pub fn get_position(&self, global: &JSGlobalObject) -> JSValue {
        BuildMessage::generate_position_object(&self.msg, global)
    }

    #[bun_jsc::host_fn(getter)]
    pub fn get_message(&self, global: &JSGlobalObject) -> JSValue {
        ZigString::init(&self.msg.data.text).to_js(global)
    }

    #[bun_jsc::host_fn(getter)]
    pub fn get_level(&self, global: &JSGlobalObject) -> JSValue {
        ZigString::init(self.msg.kind.string()).to_js(global)
    }

    pub fn finalize(this: *mut BuildMessage) {
        // SAFETY: `this` was allocated via Box::new in `create` and handed to the
        // C++ JSCell wrapper as m_ctx; finalize is called exactly once on the
        // mutator thread during lazy sweep. Dropping the Box runs `msg`'s Drop.
        unsafe { drop(Box::from_raw(this)); }
    }
}

// ──────────────────────────────────────────────────────────────────────────
// PORT STATUS
//   source:     src/jsc/BuildMessage.zig (203 lines)
//   confidence: medium
//   todos:      1
//   notes:      .classes.ts m_ctx payload; allocator field dropped; to_string_fn external-value path leaks Vec for JSC finalizer to free
// ──────────────────────────────────────────────────────────────────────────
